Fast Forward/Rewind Function

albertyeh199909 opened the pull request:
<!-- Hey there. Thank you so much for improving OuterTune, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

### What is it?

- [x] New feature (user facing)
- [ ] Update to existing feature (user facing)
- [ ] Bugfix (user facing)
- [ ] Codebase improvements or refactors (dev facing)
- [ ] Other

### Description of the changes in your PR

<!-- While bullet points are the norm in this section, feel free to write free-form text instead of a list -->

- Added buttons to seek forward/backward
- Added settings and related strings/enums to change seeking increments 5/10/15/20
- Changed auto skip next when error setting icon to skip next instead of fast forward

### Before/After Screenshots/Screen Record

<!-- If your PR changes the app's UI in any way, please include screenshots or a video showing exactly what changed, so that developers and users can pinpoint it easily. Delete this if it doesn't apply to your PR.-->

- Before:
<img width="1061" height="1918" alt="image" src="https://github.com/user-attachments/assets/0157db42-999a-46b3-8ee6-73cbe0c1e1f5" />


- After:<img width="355" height="783" alt="image" src="https://github.com/user-attachments/assets/fee2cd45-f734-4c83-8179-363ba467959f" />


Uploading OutertuneAddSeekForwardBackward.mp4…



### Fixes the following issue(s)

<!-- Prefix issues with "Fixes" so that GitHub closes them when the PR is merged (ex: "Fixes #69". Note that each "Fixes #" should be in its own item). Also add any other relevant links. -->

- Fixes #641 

### Relies on the following changes

<!-- Tag any pull requests that are required before this can be merged.
Delete this if it doesn't apply to your PR. -->

-

## Due diligence

<!-- Please mark WIP pull requests and "Draft" and only "Ready for review" once it is ready to be merged  -->
I read the contribution guidelines but I am a firsttimer. Please forgive me if I misunderstood the rules.
- [x] I have read and agreed to the [contribution guidelines](https://github.com/OuterTune/OuterTune/blob/dev/CONTRIBUTING.md).

### Merging strategy / Merge conflict resolution

Select only ONE. If you select none, or both, the first selection will used as your final choice

- [x] When merging this pull request, or in the event of merge conflicts, I give permission for the merger to rebase,
  or squash my code, or apply merge conflict amendments as needed
- [ ] When merging this pull request, or in the event of merge conflicts, I ***DO NOT*** give permission for the
  merger to modify my code to solve merge conflicts. I understand in the event of a merge conflict, I will be
  responsible to resolve merge conflicts in a way that adheres to
  the [contribution guidelines](https://github.com/OuterTune/OuterTune/blob/dev/CONTRIBUTING.md)

<!-- This pull request template is based on Newpipe's:  https://github.com/TeamNewPipe/NewPipe/ -->

albertyeh199909 committed:
player: added seek forward and backward buttons and settings

albertyeh199909 committed:
player: added seek forward and backward buttons and settings

albertyeh199909 committed:
player: added seek forward and backward buttons and settings

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/player/Player.kt:
```diff
@@ -579,6 +601,22 @@ fun BottomSheetPlayer(
+                            .align(Alignment.Center),
+                        color = onBackgroundColor,
+                        onClick = {
+                            //ExoPlayer seek increment can only be set in builder
+                            //playerConnection.player.seekForward()
```
Did you mean to delete this?

albertyeh199909 replied:
I added this in case future devs try to use this function. seekForward() uses a value that can only be set by a setter in the builder, cant be changed dynamically once the instance is created. I can delete it if you want.

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/settings/fragments/PlayerFrag.kt:
```diff
@@ -118,7 +135,7 @@ fun PlaybackBehaviourFrag() {
         title = { Text(stringResource(R.string.auto_skip_next_on_error)) },
         description = stringResource(R.string.auto_skip_next_on_error_desc),
-        icon = { Icon(Icons.Rounded.FastForward, null) },
+        icon = { Icon(Icons.Rounded.SkipNext, null) },
```
why

albertyeh199909 replied:
Hi, I changed the icon, since the original icon used Fastforward instead of skipping to the next song icon. I want to use the fastforward icon for my own setting that I added. 
Doesnt auto_skip_on_error using a skip icon make more sense?

mikooomich replied:
I'll have to revisit that one. There was a reason I used fast forward but I don't remember exactly why

albertyeh199909 replied:
> I'll have to revisit that one. There was a reason I used fast forward but I don't remember exactly why

Gotcha. Let me know if you remember or want me to undo this.

mikooomich commented on app/src/main/res/values/strings.xml:
```diff
@@ -283,6 +283,7 @@ We recycle translations from upstream, they are pulled periodically. -->
     <string name="audio_normalization">Audio normalization</string>
     <string name="auto_skip_next_on_error">Auto skip to next song when error occurs</string>
     <string name="auto_skip_next_on_error_desc">Ensure your continuous playback experience</string>
+    <string name="seek_increment">Seek increment</string>
```
outertune strings to into strings-ot

albertyeh199909 replied:
My bad. I missed that part in the guidelines. I'll fix it right away.

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/player/Player.kt:
```diff
@@ -548,6 +557,19 @@ fun BottomSheetPlayer(
+                            .align(Alignment.Center),
+                        color = onBackgroundColor,
+                        onClick = {
+                            playerConnection.player.seekTo(playerConnection.player.currentPosition - seekIncrement.value * 1000)
```
SeekIncrement should probably be in milliseconds rather than * 1000

albertyeh199909 replied:
Gotcha, that makes sense

mikooomich reviewed (changes requested):
Hello! Thank you for your PR, and sorry it took so long for me to take a look at it

As mentioned before, this should be disabled if the user so chooses (please also disable it by default). You can add a 0 or disabled enum to SeekIncrement ig

albertyeh199909 commented:
> Hello! Thank you for your PR, and sorry it took so long for me to take a look at it
> 
> As mentioned before, this should be disabled if the user so chooses (please also disable it by default). You can add a 0 or disabled enum to SeekIncrement ig


I'll make it disabled by default and add the options. How do I avoid confusing the users when they see the button is disabled and are not immediately sure how to use them? Should I add some sort of prompt or hint? Thanks for responding I was getting worried haha.

mikooomich commented:
> How do I avoid confusing the users when they see the button is disabled and are not immediately sure how to use them?

Hide if disabled.


I forgot to mention this but there's player controls on the queue. Up to you if you want to add it there too, but code wise the follow very similar patterns as the main player screen

albertyeh199909 committed:
player: added seek forward and backward buttons and settings

albertyeh199909 commented:
> > How do I avoid confusing the users when they see the button is disabled and are not immediately sure how to use them?
> 
> Hide if disabled.
> 
> I forgot to mention this but there's player controls on the queue. Up to you if you want to add it there too, but code wise the follow very similar patterns as the main player screen

Added option to hide the buttons when disabled is selected in settings. Also added the buttons in the queue player controls and moved the strings to strings-ot.xml.

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/settings/fragments/PlayerFrag.kt:
```diff
@@ -36,13 +39,30 @@ import com.dd3boh.outertune.utils.rememberPreference
+        icon = { Icon(Icons.Rounded.FastForward, null) },
+        selectedValue = seekIncrement,
+        onValueSelected = onSeekIncrementChange,
+        valueText = {
+            when (it) {
+                SeekIncrement.DISABLED -> stringResource(R.string.seek_increment_disabled)
+                else -> it.second.toString()
+            }
```
You might be able to do something like the contributor type here. https://github.com/OuterTune/OuterTune/blob/b739e81768285f36cb50dc219ccfda3d86cfff18/app/src/main/java/com/dd3boh/outertune/ui/component/ContributorCard.kt#L126

albertyeh199909 replied:
I added something similar, and move these lines into a function in the enum. Did I understand that correctly?

mikooomich commented on app/src/main/res/values/strings-ot.xml:
```diff
@@ -155,6 +155,8 @@ while also allowing users to translate our app as well. New, specific strings to
     <string name="swipe2Queue">Swipe to queue</string>
     <string name="swipe2Queue_description">Swipe on a song to add it to "enqueue next" directly after currently playing song</string>
+    <string name="seek_increment">Seek increment</string>
+    <string name="seek_increment_disabled">Disabled</string>
```
We can use the "off" string in place of disabled tbh (androidx.compose.ui.R.string.state_off, or even R.string.dark_theme_off lol)

albertyeh199909 replied:
Changed the enum name to OFF and used androidx.compose.ui.R.string.state_off. Removed the string resource I added.

mikooomich reviewed (changes requested):
A couple of minor changes, otherwise looks good to me!

albertyeh199909 committed:
player: added seek forward and backward buttons and settings
moved seekincrement setting string function into enum itself, used state off resource string instead and removed the one I added

mikooomich committed:
meow
